array_api_tests/test_operators_and_elementwise_functions.py

```diff
@@ -25,67 +15,215 @@
 from . import pytest_helpers as ph
 from . import shape_helpers as sh
 from . import xps
-from .algos import broadcast_shapes
-from .typing import Array, DataType, Param, Scalar
+from .typing import Array, DataType, Param, Scalar, ScalarType, Shape
 
 pytestmark = pytest.mark.ci
 
+
+def all_integer_dtypes() -> st.SearchStrategy[DataType]:
+    """Returns a strategy for signed and unsigned integer dtype objects."""
+    return xps.unsigned_integer_dtypes() | xps.integer_dtypes()
+
+
+def boolean_and_all_integer_dtypes() -> st.SearchStrategy[DataType]:
+    """Returns a strategy for boolean and all integer dtype objects."""
+    return xps.boolean_dtypes() | all_integer_dtypes()
+
+
+def mock_int_dtype(n: int, dtype: DataType) -> int:
+    """Returns equivalent of `n` that mocks `dtype` behaviour."""
+    nbits = dh.dtype_nbits[dtype]
+    mask = (1 << nbits) - 1
+    n &= mask
+    if dh.dtype_signed[dtype]:
+        highest_bit = 1 << (nbits - 1)
+        if n & highest_bit:
+            n = -((~n & mask) + 1)
+    return n
+
+
+# This module tests elementwise functions/operators against a reference
+# implementation. We iterate through the input array(s) and resulting array,
+# casting the indexed arrays to Python scalars and calculating the expected
+# output with `refimpl` function.
+#
+# This is finicky to refactor, but possible and ultimately worthwhile - hence
+# why these *_assert_again_refimpl() utilities exist.
+#
+# Values which are special-cased are generated and passed, but are filtered by
+# the `filter_` callable before they can be asserted against `refimpl`. We
+# automatically generate tests for special cases in the special_cases/ dir. We
+# still pass them here so as to ensure their presence doesn't affect the outputs
+# respective to non-special-cased elements.
+#
+# By default, results are casted to scalars the same way that the inputs are.
+# You can specify a cast via `res_stype, i.e. when a function accepts numerical
+# inputs but returns boolean arrays.
+#
+# By default, floating-point functions/methods are loosely asserted against. Use
+# `strict_check=True` when they should be strictly asserted against, i.e.
+# when a function should return intergrals.
+
+
+def isclose(a: float, b: float, rel_tol: float = 0.25, abs_tol: float = 1) -> bool:
```

Comment:
I guess since we are testing one float at a time, just using math.isclose here is fine. In general though, we will want an isclose, and corresponding assert_approximately_equal method that only use the array API methods so that they can be called on arrays (I will need this for the linalg tests).

I don't *think* it will be an issue, but if we do run into issues with float32 and the fact that Python `float` is 64-bit only, we may need to make this use array operations too. That would also give us more control over the exact formula used for "closeness", though I doubt that will be an issue either given our very loose tolerances.

Reply:
> In general though, we will want an isclose, and corresponding assert_approximately_equal method that only use the array API methods so that they can be called on arrays (I will need this for the linalg tests).

The `*_assert_against_refimpl` utils do this along an array/arrays, only using the Array API endpoints of indexing and scalar casting.

> I don't think it will be an issue, but if we do run into issues with float32 and the fact that Python float is 64-bit only, we may need to make this use array operations too.

Ah I see, good to keep in mind.